compiler/noirc_evaluator/src/ssa/ir/list.rs

```diff
@@ -0,0 +1,187 @@
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+/// A shared linked list type intended to be cloned
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct List<T> {
+    head: Arc<Node<T>>,
+    len: usize,
+}
+
+#[derive(Default, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+enum Node<T> {
+    #[default]
+    Nil,
+    Cons(T, Arc<Node<T>>),
+}
+
+impl<T> Default for List<T> {
+    fn default() -> Self {
+        List { head: Arc::new(Node::Nil), len: 0 }
+    }
+}
+
+impl<T> List<T> {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// This is actually a push_front since we just create a new head for the
+    /// list. This is done so that the tail of the list can still be shared.
+    /// In the case of call stacks, the last node will be main, while the top
+    /// of the call stack will be the head of this list.
+    pub fn push_back(&mut self, value: T) {
+        self.len += 1;
+        self.head = Arc::new(Node::Cons(value, self.head.clone()));
+    }
+
+    /// It is more efficient to iterate from the head of the list towards the tail.
+    /// For callstacks this means from the top of the call stack towards main.
+    fn iter_rev(&self) -> IterRev<T> {
+        IterRev { head: &self.head, len: self.len }
+    }
+
+    pub fn clear(&mut self) {
+        *self = Self::default();
+    }
+
+    pub fn append(&mut self, other: Self)
+    where
+        T: Copy + std::fmt::Debug,
+    {
+        let other = other.into_iter().collect::<Vec<_>>();
+
+        for item in other {
+            self.push_back(item);
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    fn pop_front(&mut self) -> Option<T>
```

Comment:
Both `push_back` and `pop_front` operate on the head; I understand what you meant by `push_back` actually being a "push_front", but wouldn't it be better to name both of these head operations to be the same? I thought `pop_front` will go to the opposite end from `push_back` in an O(n) operation. Why not just `push` and `pop`?

Reply:
Yeah, that's my mistake. They were originally both `*_back` to make it more easy for me to migrate existing code, then I changed `pop_front` to be the front to be more accurate but thought renaming the public `push_back` to front as well could be confusing to CallStack users who want to add something to the conceptual "back" of the structure, unaware that the structure is essentially reversed internally.

I'd be for renaming it back to pop_back.